File: 016_3SumClosest.py

class question(object):
    def threeSumClosest(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """

        if len(nums)< 3:
            print'the list is incorrect!'
            return None

        nums.sort()

        min_gap = float("inf")
        sum_r = 0
        for i in range(len(nums)):

            right = len(nums)-1
            left = i+1
            while left < right:
                tmp_sum = nums[right]+nums[left]+nums[i]
                if tmp_sum == target:
                    return target
                elif tmp_sum< target:
                    if abs(tmp_sum-target)<min_gap:
                        sum_r = tmp_sum
                        min_gap = abs(tmp_sum-target)
                    left += 1
                else:
                    if abs(tmp_sum-target)<min_gap:
                        sum_r = tmp_sum
                        min_gap = abs(tmp_sum-target)
                    right -= 1

        return sum_r

if __name__ == '__main__':

    inputs = [-1, 0, 1, 2, -1, -4]
    k = question()
    print 'Input :', inputs
    r = k.threeSumClosest(inputs,0)
    print  'Result : ',r

File: 016_3Sum_Closest.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var threeSumClosest = function(nums, target) {
    var len = nums.length;
    if (len<3){
        return(null)
    };

    nums.sort(function(a,b){return(a-b)});
    var mix_gap = Infinity;
    sum_cur = null
    var f= 0;
    while (f<len-2){
        var l= f+1,r=len-1;
        if(f>0 && nums[f]==nums[f-1]){
            f++;
            continue;
        };
        while(l<r){        
            tmp_sum = nums[f]+nums[l]+nums[r];
            if (tmp_sum=== target){
                return(target);
            } else if (tmp_sum > target){
                if(tmp_sum-target < mix_gap){
                    mix_gap = tmp_sum-target;
                    sum_cur = tmp_sum;
                };
                r--
                while(l<r && nums[r]===nums[r+1]){
                    r--
                };   
            }else{
                if(target - tmp_sum < mix_gap){
                    mix_gap = target - tmp_sum;
                    sum_cur = tmp_sum;
                };
                l++
                while(l<r && nums[l]===nums[l-1]){
                    l++;
                }; 
            };
        };
        f++;
    };
    return(sum_cur)
};

var string_example =[-1, 2, 1, -4];
var target = 1;
console.log(threeSumClosest(string_example,target));